Allow projects to link to multiple style databases, show all symbols and entities from those databases

nyalldawson opened the pull request:
This new feature allows users to set a list of QGIS style databases for a project. The style paths can point to style .db databases or .xml exports, located e.g. on a shared folder or alongside a project. When that project is loaded, symbol and label style selectors will show ALL symbols from ALL styles linked to the project. Additionally, every project now has a project-specific style database (which is stored alongside the project), for storing project-specific symbols, text formats, etc.

This has a few use cases:

1. A style database can be created eg for a single client, containing all the symbols used for that particular client and which are nice to share across different projects, but ONLY projects for that client (and you don't want to clutter up your default style database with those client-specific styles)
2. Multi user environments could store project specific styles in a shared location, and link a project to these styles. Then ALL users of the project will have the common symbology available to them, without having to manually import/update their local style database.
3. By allowing symbols to be stored in a project file, it avoids users cluttering up their global style database with project-specific symbology. It also ensures that other users of this project will also have all the required symbology available for immediate re-use when loading that project.

**This functionality requires a QGIS build based on Qt 5.13 or later, and is hidden on older builds (e.g. on Ubuntu 20.04)**

Sponsored by the QGIS Swiss user group.

Screencasts:

Saving a symbol into the project database:

https://user-images.githubusercontent.com/1829991/168707911-b6867ca4-8b4f-4afe-87e8-ce7903761db6.mp4

Adding additional shared style libraries to a project:

https://user-images.githubusercontent.com/1829991/168708342-085c5009-4c1c-4fbe-8e58-92dcbb3e4557.mp4

nyalldawson committed:
Deprecate QgsStyle::save, which is a no-op call as all the functionality
is ifdefed out and does nothing

nyalldawson committed:
Documentation, const

nyalldawson committed:
Don't automatically change the filename for a QgsStyle when exporting
or importing XML files

The file name should always represent the original source (e.g.
style .db file or .xml file if loaded initially from a .xml file), NOT
the file it's been most recently exported to!

nyalldawson committed:
Fix filename handling for QgsStyle

nyalldawson committed:
Expose style name and filename to models

nyalldawson committed:
Rename Project Properties 'Default styles' tab to just 'Styles'

nyalldawson committed:
Add methods to set style database paths to QgsProjectStyleSettings

nyalldawson committed:
Allow parent object for QgsStyle

nyalldawson committed:
Profile loading style properties

nyalldawson committed:
Remove need to manually call QgsStyle::upgradeIfRequired

nyalldawson committed:
Load project associated style databases

nyalldawson committed:
add combined model

nyalldawson committed:
Add method to remove style from combined model

nyalldawson committed:
Add combined style model to project

nyalldawson committed:
Allow project style paths to be xml exports of databases

nyalldawson committed:
Add a way for QgsSingleItemModel to have multiple column items

nyalldawson committed:
Combined model should have two columns

nyalldawson committed:
Fix header data for combined model

nyalldawson committed:
Ensure filename is set for project styles

nyalldawson committed:
Allow retrieving matching project style by path

nyalldawson committed:
Add entity name to style model roles

nyalldawson committed:
[feature] Show entities from all project style databases in symbol and label
selectors

nyalldawson commented:
Also should add: this requires a QGIS build based on Qt 5.13 or greater only

nirvn commented:
My two cents: +1 on freeze exception. It'd go along very well with the default style revamp done last week and as pointed out does fill an important gap with our style handling in the context of multi user environment.

nyalldawson committed:
Move QgsProject enums to Qgis, and add new read flag to skip style loading

Ensure that there's no speed regression when loading projects for
server and other special cases

nyalldawson committed:
Add missing files

nyalldawson committed:
Fix build

nyalldawson committed:
Dox

nyalldawson committed:
Add setting in project properties to configure linked style databases
for project

nyalldawson committed:
Add nice heading styling for combined style models

nyalldawson committed:
Don't always call QgsStyle::upgradeIfRequired when loading a QgsStyle
.db file

Otherwise the database may get modified and a bunch of default symbols
added to it. We should only EVER call upgradeIfRequired on the default
style database, so move the call to a more suitable place.

nyalldawson committed:
Add button to create new empty style database in project properties

nyalldawson committed:
Allow lazy setting of style manager QgsStyle

nyalldawson committed:
Add model for project style databases

nyalldawson committed:
Show combo box for selecting style database from project styles in style manager dialog

nyalldawson committed:
Add buttons for loading and creating styles to Style Manager dialog too

nyalldawson committed:
Keep filters when style is changed

nyalldawson committed:
Hide non-functional database switch on old Qt

nyalldawson committed:
Allow choice of style database when saving styles

nyalldawson committed:
Add path role to model

nyalldawson committed:
Don't offer to save new symbols in read only style databases

nyalldawson committed:
Fix doxygen warnings

nyalldawson committed:
Better way to mark styles as read only

nyalldawson committed:
Ensure style manager dialog responds correctly to read only styles

nyalldawson committed:
Cleanup handling of default label format for project, and avoid a QgsProject::instance() use in core

nyalldawson committed:
Cleanup some more of the default style/project instance initialisation spaghetti

And avoid some more QgsProject::instance() in core

nyalldawson committed:
Fix test

nyalldawson committed:
Add project style database, which contains styles specific to each
individual project.

Project style database is embedded in the .qgz archive file

nyalldawson committed:
Indentation

nirvn commented on src/gui/symbology/qgsstylemanagerdialog.cpp:
```diff
@@ -213,49 +251,26 @@ QgsStyleManagerDialog::QgsStyleManagerDialog( QgsStyle *style, QWidget *parent,
-    btnAddSmartgroup->setEnabled( false );
+    // only show add item if the btn doesn't have a menu -- otherwise it should show the menu instead!
+    if ( !btnAddItem->menu() )
+    { addItem(); }
```
Ew formatting :wink: we gotta tame astyle to handle this better one day :)

nirvn commented on src/gui/symbology/qgsstylemanagerdialog.cpp:
```diff
@@ -518,18 +460,166 @@ QgsStyleManagerDialog::QgsStyleManagerDialog( QgsStyle *style, QWidget *parent,
+  {
+    if ( !mActionCopyToDefault )
+    {
+      mActionCopyToDefault = new QAction( tr( "Copy Selection to Default Style…" ), this );
```
Would it be useful to have a Copy Selection to Project Style when the current style isn't the project style?

nyalldawson replied:
I'll defer this to a follow up

nirvn reviewed (approved):
Reviewed, looking good.

nyalldawson committed:
Update test

nyalldawson committed:
Improve indentation
